cli/src/utils.rs

```diff
@@ -16,30 +16,98 @@ pub struct FileOrUrl {
     /// The path to the encoded metadata file.
     #[clap(long, value_parser)]
     file: Option<PathBuf>,
+    /// Specify the metadata version.
+    ///
+    ///  - unstable:
+    ///
+    ///    Use the latest unstable metadata of the node.
+    ///
+    ///  - number
+    ///
+    ///    Use this specific metadata version.
+    ///
+    /// Defaults to 14.
+    #[clap(long)]
+    version: Option<MetadataVersion>,
 }
 
 impl FileOrUrl {
     /// Fetch the metadata bytes.
     pub async fn fetch(&self) -> color_eyre::Result<Vec<u8>> {
-        match (&self.file, &self.url) {
+        match (&self.file, &self.url, &self.version) {
             // Can't provide both --file and --url
-            (Some(_), Some(_)) => {
+            (Some(_), Some(_), _) => {
                 eyre::bail!("specify one of `--url` or `--file` but not both")
             }
             // Load from --file path
-            (Some(path), None) => {
+            (Some(path), None, None) => {
                 let mut file = fs::File::open(path)?;
                 let mut bytes = Vec::new();
                 file.read_to_end(&mut bytes)?;
                 Ok(bytes)
             }
+            // Cannot load the metadata from the file and specify a version to fetch.
+            (Some(_), None, Some(_)) => {
+                // Note: we could provide the ability to convert between metadata versions
+                // but that would be involved because we'd need to convert
+                // from each metadata to the latest one and from the
+                // latest one to each metadata version. For now, disable the conversion.
+                eyre::bail!("`--file` is incompatible with `--version`")
+            }
             // Fetch from --url
-            (None, Some(uri)) => Ok(subxt_codegen::utils::fetch_metadata_bytes(uri).await?),
+            (None, Some(uri), version) => Ok(subxt_codegen::utils::fetch_metadata_bytes(
+                uri,
+                version.map(Into::into).unwrap_or_default(),
+            )
+            .await?),
             // Default if neither is provided; fetch from local url
-            (None, None) => {
+            (None, None, version) => {
                 let uri = Uri::from_static("http://localhost:9933");
-                Ok(subxt_codegen::utils::fetch_metadata_bytes(&uri).await?)
+                Ok(subxt_codegen::utils::fetch_metadata_bytes(
+                    &uri,
+                    version.map(Into::into).unwrap_or_default(),
+                )
+                .await?)
+            }
+        }
+    }
+}
+
+/// The metadata version to fetch from the node.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct MetadataVersion {
+    version: u32,
+}
+
+impl std::str::FromStr for MetadataVersion {
+    type Err = String;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        const SUPPORTED_VERSIONS: &[u32] = &[14];
```

Comment:
Personally I don't think I'd have this; I'd just try fetching whatever version they ask for (or checking the list of versions the node provides) and returning an error if the version isn't available on that node.

Subxt itself may only support certain versions, but the macro/codegen can be the thing that complains if the user asks for a version that's not supported by Subxt, and the CLI tool can just do it's best to download whatever is asked for from a node :)